Move Tensor.grad back into C++

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74728
* #74727
* __->__ #76675

`Tensor.grad` was moved to python in #30531 to add a warning. However,
that warning has since been lowered into C++ so this wrapper is no
longer necessary.

peterbell10 committed:
Move Tensor.grad back into C++

`Tensor.grad` was moved to python in #30531 to add a warning. However,
that warning has since been lowered into C++ so this wrapper is no
longer necessary.

[ghstack-poisoned]

peterbell10 committed:
Update on "Move Tensor.grad back into C++"


`Tensor.grad` was moved to python in #30531 to add a warning. However,
that warning has since been lowered into C++ so this wrapper is no
longer necessary.

[ghstack-poisoned]

albanD reviewed (approved):
Nice cleanup, thanks!

peterbell10 committed:
Update on "Move Tensor.grad back into C++"


`Tensor.grad` was moved to python in #30531 to add a warning. However,
that warning has since been lowered into C++ so this wrapper is no
longer necessary.

[ghstack-poisoned]

peterbell10 committed:
Update on "Move Tensor.grad back into C++"


`Tensor.grad` was moved to python in #30531 to add a warning. However,
that warning has since been lowered into C++ so this wrapper is no
longer necessary.

[ghstack-poisoned]

peterbell10 committed:
Rebase and fix merge conflicts on "Move Tensor.grad back into C++"


`Tensor.grad` was moved to python in #30531 to add a warning. However,
that warning has since been lowered into C++ so this wrapper is no
longer necessary.

[ghstack-poisoned]

peterbell10 committed:
Update on "Move Tensor.grad back into C++"


`Tensor.grad` was moved to python in #30531 to add a warning. However,
that warning has since been lowered into C++ so this wrapper is no
longer necessary.

[ghstack-poisoned]

peterbell10 committed:
Update on "Move Tensor.grad back into C++"


`Tensor.grad` was moved to python in #30531 to add a warning. However,
that warning has since been lowered into C++ so this wrapper is no
longer necessary.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge this please

albanD commented:
@pytorchbot revert this please as this broke master lint: https://github.com/pytorch/pytorch/runs/6358762016?check_suite_focus=true

peterbell10 committed:
Update on "Move Tensor.grad back into C++"


`Tensor.grad` was moved to python in #30531 to add a warning. However,
that warning has since been lowered into C++ so this wrapper is no
longer necessary.

[ghstack-poisoned]

peterbell10 committed:
Update on "Move Tensor.grad back into C++"


`Tensor.grad` was moved to python in #30531 to add a warning. However,
that warning has since been lowered into C++ so this wrapper is no
longer necessary.

[ghstack-poisoned]

peterbell10 commented on torch/nn/utils/clip_grad.py:
```diff
@@ -29,17 +29,17 @@ def clip_grad_norm_(
     if isinstance(parameters, torch.Tensor):
         parameters = [parameters]
-    parameters = [p for p in parameters if p.grad is not None]
+    grads = [p.grad for p in parameters if p.grad is not None]
```
@albanD the lint failures where cases where because `mypy` now has typing information for `.grad` and needs to be convinced that the value won't be `None`.

albanD replied:
Ho and `mypy` doesn't see the `is not None` check?

peterbell10 replied:
I think the type refinement maybe only works within a single scope, so once you leave the list comprehension it forgets that `p.grad` won't be none. However this new comprehension results in `grads` being non-optional Tensors so it does work.

malfet commented:
@peterbell10 adding `ciflow/trunk` would not help here (i.e. master and PR lint rules are the same), you just need to rebase your PR against latest master and try again

peterbell10 committed:
Rebase on "Move Tensor.grad back into C++"


`Tensor.grad` was moved to python in #30531 to add a warning. However,
that warning has since been lowered into C++ so this wrapper is no
longer necessary.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
